### PR TITLE
Wordlist export features

### PR DESCRIPTION
**Pull request by irina060981**
For the issue #200 

1. Added new fields for worditem
    - updatedDT
    - frequency

2. Added additional filter
    - most recent (by updatedDT)
    - most often (by frequency)

    
image

3. each additional filter has filterAmount field, it defines amount of words in filter

image

    it gets default data from `featureOptioins.wordlistFilterAmountDefault`

4. added another `featureOptions.wordlistMaxFlashcardExport` option
    used inside download confirmation block
    is saved to user profile settings if changed

    
image

5. On flashcard export if worditem doesn't have shortDefs - it inits additional lexicalQuery
    and download the wordlist only after all lexical requests are complete
    while lexical data is loading - progressbar is shown
    
image

6. also added `frequency` and `updatedDT` to wordlist ui and added sorting for them

### CODE REVIEW

**Comment by irina060981** [comment]
will update tests later, forgot about them
anyway - @balmas , would be glad to get some feedback!

**Comment by balmas** [comment]
@irina060981 I think this is a great start ! Thanks for trying to make sense of the requirements :-)

a few thoughts on the UI:

(1) I don't think I want it to be so easy for the user to change the max items for download, since this something we are restricting to prevent performance issues.  So rather than offering them the option to change the max on the download dialog, I'd like to just show them a notice, e.g. something like "Note: the Flashcard export is limited to a maximum of XX words." where XX is taken from the setting.  And then we can put the option to change the setting on the Features option tab.

(2) I'm thinking about these 2 levels of limits and I think they are a little confusing and I'm not sure it would make sense for them to be used in conjunction.  Can we just put the new limits in the existing box? And then precede the input box for the `wordlistFilterAmountDefault` value with the word "Top" ? (I expect we will have to have a few iterations on the UI but I think I'd like to start with that).

(3) the a/z sort icons need to be different for the new fields, because the alphas don't make sense there. Can we just have up/down arrows?

(4) with the additional fields, we probably need to add column headers to the display. I don't think it's obvious what the number and the date mean.

Can you make those changes and then I'll take a closer look at the code?

Thanks!

**Comment by irina060981** [comment]
Updated ui
image

image

File: packages/components/src/vue/vuex-modules/data/lexis.js [hunk 1]
```diff
@@ -200,7 +200,10 @@ export default class Lexis extends Module {
     }
     if (!wordUsageExamples) { wordUsageExamples = this._appApi.getWordUsageExamplesQueryParams(textSelector) }
 
-    this._appApi.newLexicalRequest(textSelector.normalizedText, textSelector.languageID, textSelector.data, source)
+    if (!forWordlist) {
+      this._appApi.newLexicalRequest(textSelector.normalizedText, textSelector.languageID, textSelector.data, source)
+    }
```
Comment (balmas) [inline packages/components/src/vue/vuex-modules/data/lexis.js:205]: I'm wondering if there are going to be any unexpected side effects of issuing the lexicalQuery.getData request without going through the regular process of initializing a new lexical request. getData issues a number of events that affect the interface.  It might be that we need to do more here, and suppress events during the lexical query getData when we're doing this sort of background request. I'd like to get some thoughts from @kirlat on this.

File: packages/components/tests/vue/components/word-list/word-language-panel.test.js [hunk 2]
```diff
@@ -594,7 +608,7 @@ describe('word-language-panel.test.js', () => {
     expect(result[2].targetWord).toEqual('cupidinibus')
   })
 
-  it('26 WordLanguagePanel - method downloadList executes Download.collectionToCSV, Download.downloadBlob and hides downloadAll confirmation box', () => {
+  it.skip('26 WordLanguagePanel - method downloadList executes Download.collectionToCSV, Download.downloadBlob and hides downloadAll confirmation box', () => {
```
Comment (balmas) [inline packages/components/tests/vue/components/word-list/word-language-panel.test.js:611]: why are we skipping this test now?

File: packages/wordlist/src/controllers/user-data-manager.js [hunk 3]
```diff
@@ -110,6 +110,7 @@ export default class UserDataManager {
     }
     try {
       params.source = params.source||'both'
+      // params.source = 'local'
```
Comment (balmas) [inline packages/wordlist/src/controllers/user-data-manager.js:113]: remove this and below?

File: packages/wordlist/src/storage/remote-db-adapter.js [hunk 4]
```diff
@@ -172,6 +171,9 @@ export default class RemoteDBAdapter {
           this.errors.push(error)
         }
       }
+      if (error.message === 'Request failed with status code 401') {
+        return []
+      }
```
Comment (balmas) [inline packages/wordlist/src/storage/remote-db-adapter.js:176]: why the change in behavior here?

File: packages/wordlist/src/storage/worditem-remotedb-driver.js [hunk 5]
```diff
@@ -183,6 +186,8 @@ export default class WordItemRemoteDbDriver {
     } else {
       result.context = []
     }
+    // console.info('remote _serialize wordItem', wordItem.frequency, wordItem)
```
Comment (balmas) [inline packages/wordlist/src/storage/worditem-remotedb-driver.js:189]: remove debugging

File: packages/components/src/settings/feature-options-defaults.json [hunk 6]
```diff
@@ -233,6 +233,22 @@
         }
       ],
       "limitByLangs": ["lat", "grc"]
+    },
+    "wordlistMaxFlashcardExport": {
+      "defaultValue": 25,
+      "labelText": "Wordlist Flashcard Max export",
+      "number": true,
+      "minValue": 1,
+      "maxValue": 5000,
```
[the remark this answers is not in the document]
Comment (irina060981) [reply packages/components/src/settings/feature-options-defaults.json:242]: fixed

File: packages/components/src/settings/feature-options-defaults.json [hunk 7]
```diff
@@ -233,6 +233,22 @@
...
+      "values": []
+    },
+    "wordlistFilterAmountDefault": {
+      "defaultValue": 10,
+      "labelText": "Wordlist Filter Amount default",
```
[the remark this answers is not in the document]
Comment (irina060981) [reply packages/components/src/settings/feature-options-defaults.json]: done

File: packages/components/src/settings/feature-options-defaults.json [hunk 8]
```diff
@@ -233,6 +233,22 @@
...
+      "number": true,
+      "minValue": 1,
+      "maxValue": 25,
```
[the remark this answers is not in the document]
Comment (irina060981) [reply packages/components/src/settings/feature-options-defaults.json]: done

File: packages/components/src/locales/en-us/messages-word-list.json [hunk 9]
```diff
@@ -144,5 +144,26 @@
...
         "description": "Download list for flashcards",
         "component": "WordLanguagePanel"
+    },
+    "WORDLIST_FILTER_BYMOSTRECENT": {
+        "message": "by most recent",
+        "description": "Filter by max amount with sorting by updatedDT desc",
+        "component": "WordFilterPanel"
+    },
+    "WORDLIST_FILTER_AMOUNT_PLACEHOLDER": {
+        "message": "top amount",
```
[the remark this answers is not in the document]
Comment (irina060981) [reply packages/components/src/locales/en-us/messages-word-list.json]: done

File: packages/components/src/vue/components/word-list/download-confirmation.vue [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+        Download.downloadBlob(result, `wordlist-${this.languageCode}.csv`)
+        this.$emit('changeShowDownloadBox', false)
+      },
+      cancelDownloadList () {
+        this.$emit('changeShowDownloadBox', false)
+      },
+      featureOptionChanged (name, value) {
+        let keyinfo = Options.parseKey(name)
+
+        console.info('featureOptionChanged - ', keyinfo.name, value)
```
[the remark this answers is not in the document]
Comment (irina060981) [reply packages/components/src/vue/components/word-list/download-confirmation.vue:185]: done

File: packages/components/src/settings/feature-options-defaults.json [hunk 11]
```diff
@@ -233,6 +233,22 @@
...
+    "wordlistMaxFlashcardExport": {
+      "defaultValue": 25,
+      "labelText": "Wordlist Flashcard Max export",
+      "number": true,
+      "minValue": 1,
+      "maxValue": 5000,
+      "values": []
+    },
+    "wordlistFilterAmountDefault": {
+      "defaultValue": 10,
```
[the remark this answers is not in the document]
Comment (irina060981) [reply packages/components/src/settings/feature-options-defaults.json:246]: done

File: packages/components/src/vue/components/word-list/word-filter-panel.vue [hunk 12]
```diff
@@ -2,50 +2,64 @@
...
-            <input v-model="textInput"  class="alpheios-input alpheios-wordlist-header-input-filterBy"
-                  :placeholder="currentTypeFilter.textInputPlaceholder"
-                  v-on:keyup.enter = "clickFilterBy"
-                  v-on:input = "filterVariants"
-                  v-on:focus = "filterVariants"
-                  v-on:blur = "hideAutocomplete"
-                  autocapitalize = "off"
-                  autocorrect = "off"
-                  >
+          <div class="alpheios-wordlist-header-select-filterBy-first">
```
[the remark this answers is not in the document]
Comment (irina060981) [reply packages/components/src/vue/components/word-list/word-filter-panel.vue:5]: no, it is a wrapper around the number filed - I have changed the class name to reduce misunderstanding

File: packages/components/src/vue/vuex-modules/data/lexis.js (hunk 1 above)
Comment (irina060981) [reply packages/components/src/vue/vuex-modules/data/lexis.js:205]: From my point of view `newLexicalRequest` method change data in ui, app, storage modules.
It is not needed for background process of updating homonym.
I belive that we need only to update homonym object here - that's why I think it is ok.
Would be glad to @kirlat 's opinion.

File: packages/components/tests/vue/components/word-list/word-language-panel.test.js (hunk 2 above)
Comment (irina060981) [reply packages/components/tests/vue/components/word-list/word-language-panel.test.js:611]: I need to update it because this features was moved to a separate component - fixed

File: packages/wordlist/src/controllers/user-data-manager.js (hunk 3 above)
Comment (irina060981) [reply packages/wordlist/src/controllers/user-data-manager.js:113]: removed

File: packages/wordlist/src/storage/remote-db-adapter.js (hunk 4 above)
Comment (irina060981) [reply packages/wordlist/src/storage/remote-db-adapter.js:176]: oh, I added it after problems with testUser - without this it was failed with unproper result
I belive that it is better to simply get empty array and of course show all errors as before including this 401 error

File: packages/wordlist/src/storage/worditem-remotedb-driver.js (hunk 5 above)
Comment (irina060981) [reply packages/wordlist/src/storage/worditem-remotedb-driver.js:189]: done

**Comment by irina060981** [comment]
@balmas , if I understood you right - I could merge it after above small fixes.
will do that :)

File: packages/components/src/vue/vuex-modules/data/lexis.js (hunk 1 above)
Comment (kirlat) [reply packages/components/src/vue/vuex-modules/data/lexis.js:205]: @irina060981 I've tried to simplify a number of options of lexical query methods and I think it is important in order to keep it maintainable. Right now we have two types of requests identified by a `source` prop: a get selected text request from a page (identified by a `page` source) and a one initiated by the lookup component (source is `lookup`). It looks to me (please correct me if I'm wrong) that this is a third type of request. So maybe we can use a `source` filed to identify it (i.e. having it set to something like `wordList`)? We can modify a `lookupText()` API method so that it will take `source` as an additional parameter and will pass it through to `lexicalQuery`. Or we can create another special method like `wordListLookup()` that will be similar to `lookupText()` but will use a source of `wordList`. What do you think about that?

File: packages/wordlist/src/storage/remote-db-adapter.js (hunk 4 above)
Comment (balmas) [reply packages/wordlist/src/storage/remote-db-adapter.js:176]: ok, thanks!

File: packages/components/src/vue/components/word-list/word-filter-panel.vue (hunk 12 above)
[the remark this answers is not in the document]
Comment (balmas) [reply packages/components/src/vue/components/word-list/word-filter-panel.vue:5]: ok thanks.
